File: app/src/main/java/com/android/mathias/velocity/FragmentCurrent.java
package com.android.mathias.velocity;

import android.animation.ObjectAnimator;
import android.app.Activity;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.BitmapFactory;
import android.graphics.Color;
import android.os.Bundle;
import android.os.Handler;
import android.os.SystemClock;
import android.preference.PreferenceManager;
import android.support.annotation.NonNull;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.NotificationCompat;
import android.text.format.DateFormat;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.DecelerateInterpolator;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.ListPopupWindow;
import android.widget.ProgressBar;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Objects;
import java.util.concurrent.TimeUnit;

public class FragmentCurrent extends android.support.v4.app.Fragment {

    private static int NOTIFICATION_ID = 1;
    private static String CHANNEL_ID = "channel_walk";
    TextView mTimeView;
    TextView mRouteView;
    TimeState mTimeState;
    long mStartTime;
    long mLastStopTime;
    Route mCurrentWalkRoute;
    ObjectAnimator mAnimator;
    NotificationManager mNotificationManager;
    SharedPreferences mSharedPref;
    NotificationChannel mChannel;
    Activity mActivity;

    FloatingActionButton mFab;
    Button mBtnR;
    Handler mHandler;
    ProgressBar mProgressBar;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        final View view = inflater.inflate(R.layout.fragment_current, container, false);
        setHasOptionsMenu(true);
        return view;
    }

    private void toggleStopwatch() {
        switch (mTimeState) {
            case STOPPED: startWalk(); break;
            case RUNNING: pauseStopwatch(); break;
            case PAUSED:  resumeStopwatch(); break;
            default: break;
        }
    }

    private void startWalk() {
        final List<String> routeNames = new ArrayList<>();
        for (Route r : DBManager.getRoutes(getContext(), null)) { routeNames.add(r.getName()); }
        String defaultRouteName = mSharedPref.getString("default_route", "None");
        if (defaultRouteName.equals("None")) { mCurrentWalkRoute = new Route("No route set"); }
        else { mCurrentWalkRoute = DBManager.getRoutes(getContext(), defaultRouteName).get(0); }
        if (routeNames.size() > 0 && defaultRouteName.equals("None")) {
            final ArrayAdapter<CharSequence> adapter = new ArrayAdapter<>(Objects.requireNonNull(getContext()), R.layout.support_simple_spinner_dropdown_item);
            adapter.addAll(routeNames);
            adapter.add("None");
            final ListPopupWindow lpw = new ListPopupWindow(getContext());
            lpw.setAdapter(adapter);
            lpw.setAnchorView(mFab);
            lpw.setContentWidth((int) (((View)mFab.getParent()).getWidth()/2.5));
            lpw.setOnItemClickListener((adapterView, view, i, l) -> {
                if (adapterView.getItemAtPosition(i).toString().equals("None")) {
                    mCurrentWalkRoute = new Route("No route set");
                } else {
                    mCurrentWalkRoute = DBManager.getRoutes(getContext(), adapterView.getItemAtPosition(i).toString()).get(0);
                }
                startStopwatch();
                lpw.dismiss();
            });
            lpw.show();
        } else {
            startStopwatch();
        }
    }

    private void startStopwatch() {
        mStartTime = SystemClock.elapsedRealtime();
        mTimeState = TimeState.RUNNING;
        updateUI();
        mHandler = new Handler();
        mHandler.post(mRunnable);
    }

    protected void pauseStopwatch() {
        mLastStopTime = SystemClock.elapsedRealtime();
        mTimeState = TimeState.PAUSED;
        updateUI();
        mHandler.removeCallbacks(mRunnable);
        buildNotification();
    }

    protected void resumeStopwatch() {
        mStartTime = mStartTime + (SystemClock.elapsedRealtime() - mLastStopTime);
        mTimeState = TimeState.RUNNING;
        updateUI();
        mHandler.post(mRunnable);
    }

    protected void stopWalk() {
        long walkTime = SystemClock.elapsedRealtime() - mStartTime;
        Walk walk = new Walk(walkTime, new Date(), mCurrentWalkRoute);
        mLastStopTime = 0;
        mCurrentWalkRoute = null;
        mTimeState = TimeState.STOPPED;
        updateUI();
        DBManager.saveWalk(getContext(), walk);
    }

    private void updateUI() {
        switch (mTimeState) {
            case STOPPED:
                mFab.setImageResource(android.R.drawable.ic_media_play);
                mBtnR.setClickable(false);
                mBtnR.setVisibility(View.INVISIBLE);
                mRouteView.setText("");
                mProgressBar.setProgress(0);
                if (mNotificationManager.getActiveNotifications() != null) mNotificationManager.cancel(NOTIFICATION_ID);
                if (mHandler != null) mHandler.removeCallbacks(mRunnable);
                break;
            case RUNNING:
                mFab.setImageResource(android.R.drawable.ic_media_pause);
                mBtnR.setClickable(true);
                mBtnR.setVisibility(View.VISIBLE);
                mRouteView.setText(mCurrentWalkRoute.getName());
                break;
            case PAUSED:
                mFab.setImageResource(android.R.drawable.ic_media_play);
                mBtnR.setClickable(true);
                mBtnR.setVisibility(View.VISIBLE);
                mTimeView.setText(DateFormat.format("mm:ss", new Date(mLastStopTime-mStartTime)));
                mRouteView.setText(mCurrentWalkRoute.getName());
                break;
            default: break;
        }
    }

    private void buildNotification() {
        NotificationCompat.Builder builder = new NotificationCompat.Builder(mActivity, CHANNEL_ID)
                .setSmallIcon(R.drawable.ic_current)
                .setContentTitle(mTimeState == TimeState.RUNNING ? "Ongoing walk" : "Walk paused")
                .setSubText(mCurrentWalkRoute.getName())
                .setContentText(DateFormat.format("mm:ss", new Date(SystemClock.elapsedRealtime()-mStartTime)))
                .setOngoing(mTimeState == TimeState.RUNNING)
                .setLargeIcon(BitmapFactory.decodeResource(getResources(), R.mipmap.ic_app))
                .setAutoCancel(true);
        Intent notificationIntent = new Intent(getActivity(), ActivityMain.class);
        notificationIntent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_SINGLE_TOP);
        builder.setContentIntent(PendingIntent.getActivity(getActivity(), 0, notificationIntent, 0));
        addNotificationActions(builder);
        mNotificationManager.notify(NOTIFICATION_ID, builder.build());
    }

    private void addNotificationActions(NotificationCompat.Builder builder) {
        switch (mTimeState) {
            case RUNNING:
                Intent pauseIntent = new Intent(getActivity(), ActivityMain.class);
                pauseIntent.setAction(getString(R.string.notification_action_pause));
                PendingIntent pausePendingIntent = PendingIntent.getActivity(getActivity(), 0, pauseIntent, 0);
                NotificationCompat.Action pauseAction = new NotificationCompat.Action.Builder(
                        R.drawable.ic_walk, getString(R.string.notification_action_pause), pausePendingIntent).build();
                builder.addAction(pauseAction);
                break;
            case PAUSED:
                Intent resumeIntent = new Intent(getActivity(), ActivityMain.class);
                resumeIntent.setAction(getString(R.string.notification_action_resume));
                PendingIntent resumePendingIntent = PendingIntent.getActivity(getActivity(), 0, resumeIntent, 0);
                NotificationCompat.Action resumeAction = new NotificationCompat.Action.Builder(
                        R.drawable.ic_walk, getString(R.string.notification_action_resume), resumePendingIntent).build();
                builder.addAction(resumeAction);
            default: break;
        }
        //no checks as far as TimeState.STOPPED has no notification
        Intent stopIntent = new Intent(getActivity(), ActivityMain.class);
        stopIntent.setAction(getString(R.string.notification_action_stop));
        PendingIntent stopPendingIntent = PendingIntent.getActivity(getActivity(), 0, stopIntent, 0);
        NotificationCompat.Action stopAction = new NotificationCompat.Action.Builder(
                R.drawable.ic_walk, getString(R.string.notification_action_stop), stopPendingIntent).build();
        builder.addAction(stopAction);
    }

    private void setProgressSeconds(int seconds) {
        if (mAnimator != null && mAnimator.isRunning()) mAnimator.cancel();
        mAnimator = ObjectAnimator.ofInt(mProgressBar, "progress", seconds*100);
        mAnimator.setDuration(1500);
        mAnimator.setInterpolator(new DecelerateInterpolator());
        mAnimator.start();
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        menu.clear();
        inflater.inflate(R.menu.menu_current, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_settings:
                startActivity(new Intent(mActivity, ActivitySettings.class));
                break;
            default: break;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onResume() {
        super.onResume();
        //initializations
        if (mActivity == null) mActivity = getActivity();
        mTimeView = Objects.requireNonNull(mActivity).findViewById(R.id.timer);
        mRouteView = mActivity.findViewById(R.id.txt_current_route);
        mProgressBar = mActivity.findViewById(R.id.progressBar);
        mFab = mActivity.findViewById(R.id.fab_current_toggle);
        mFab.setOnClickListener(view -> toggleStopwatch());
        mBtnR = mActivity.findViewById(R.id.fab_current_stop);
        mBtnR.setOnClickListener(view -> stopWalk());
        if (mTimeState == null) mTimeState = TimeState.STOPPED;
        if (mNotificationManager == null) mNotificationManager = (NotificationManager) Objects.requireNonNull(getActivity()).getSystemService(Context.NOTIFICATION_SERVICE);
        mSharedPref = PreferenceManager.getDefaultSharedPreferences(getContext());
        //custom actions
        if (mTimeState == TimeState.RUNNING) mHandler.post(mRunnable);
        if (mChannel == null) createNotificationChannel();
        updateUI();
    }

    private void createNotificationChannel() {
        mChannel = new NotificationChannel(CHANNEL_ID, getString(R.string.channel_name), NotificationManager.IMPORTANCE_LOW);
        mChannel.setDescription(getString(R.string.channel_description));
        mChannel.enableLights(mSharedPref.getBoolean("notifications_led", false));
        mChannel.setLightColor(Color.GREEN);
        mChannel.enableVibration(mSharedPref.getBoolean("notifications_vibrate", false));
        mChannel.setVibrationPattern(new long[]{100, 200, 300, 400, 500, 400, 300, 200, 400});
        mNotificationManager.createNotificationChannel(mChannel);
    }

    private enum TimeState {
        RUNNING, PAUSED, STOPPED
    }

    private Runnable mRunnable = new Runnable() {
        @Override
        public void run() {
            long time = (SystemClock.elapsedRealtime() - mStartTime);
            int sec = (int) (TimeUnit.MILLISECONDS.toSeconds(time)) % 60;
            mTimeView.setText(DateFormat.format("mm:ss", new Date(time)));
            if (mSharedPref.getBoolean("enable_notifications", false))  {
                buildNotification();
            } else if (mNotificationManager.getActiveNotifications() != null) {
                mNotificationManager.cancel(NOTIFICATION_ID);
            }
            setProgressSeconds(sec);
            mHandler.postDelayed(this, 1000);
        }
    };
}

File: app/src/main/java/com/android/mathias/velocity/FragmentCurrent.kt
package com.android.mathias.velocity

import android.animation.ObjectAnimator
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.graphics.BitmapFactory
import android.graphics.Color
import android.os.Bundle
import android.os.Handler
import android.os.SystemClock
import android.preference.PreferenceManager
import android.text.format.DateFormat
import android.view.*
import android.view.animation.DecelerateInterpolator
import android.widget.*
import androidx.core.app.NotificationCompat
import androidx.fragment.app.Fragment
import com.android.mathias.velocity.FragmentCurrent.TimeState.*
import com.android.mathias.velocity.db.DBManager
import com.android.mathias.velocity.model.Route
import com.android.mathias.velocity.model.Walk
import com.google.android.material.floatingactionbutton.FloatingActionButton
import java.util.*
import java.util.concurrent.TimeUnit

class FragmentCurrent : Fragment() {
    private lateinit var mTimeView: TextView
    private lateinit var mRouteView: TextView
    private lateinit var mFab: FloatingActionButton
    private lateinit var mBtnR: Button
    private lateinit var mProgressBar: ProgressBar

    private lateinit var mNotificationManager: NotificationManager
    private lateinit var mSharedPref: SharedPreferences
    private lateinit var mChannel: NotificationChannel

    private var mTimeState: TimeState? = null
    private var mStartTime: Long = 0
    private var mLastStopTime: Long = 0
    private var mCurrentWalkRoute: Route? = null
    private var mAnimator: ObjectAnimator? = null
    private var mHandler: Handler? = null

    private val mRunnable = object : Runnable {
        override fun run() {
            val time = SystemClock.elapsedRealtime() - mStartTime
            val sec = TimeUnit.MILLISECONDS.toSeconds(time).toInt() % 60
            mTimeView.text = DateFormat.format("mm:ss", Date(time))
            if (mSharedPref.getBoolean("enable_notifications", false)) {
                buildNotification()
            } else if (mNotificationManager.activeNotifications != null) {
                mNotificationManager.cancel(NOTIFICATION_ID)
            }
            setProgressSeconds(sec)
            mHandler!!.postDelayed(this, 1000)
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_current, container, false)
        setHasOptionsMenu(true)
        mTimeView = view.findViewById(R.id.timer)
        mRouteView = view.findViewById(R.id.txt_current_route)
        mProgressBar = view.findViewById(R.id.progressBar)
        mFab = view.findViewById(R.id.fab_current_toggle)
        mBtnR = view.findViewById(R.id.fab_current_stop)
        mFab.setOnClickListener { toggleStopwatch() }
        mBtnR.setOnClickListener { stopWalk() }
        mSharedPref = PreferenceManager.getDefaultSharedPreferences(context)
        createNotificationChannel()
        return view
    }

    private fun toggleStopwatch() {
        when (mTimeState) {
            STOPPED -> startWalk()
            RUNNING -> pauseStopwatch()
            PAUSED -> resumeStopwatch()
        }
    }

    private fun startWalk() {
        val routeNames = ArrayList<String>()
        for (r in DBManager.getRoutes(context!!, null)) {
            routeNames.add(r.name!!)
        }
        val defaultRouteName = mSharedPref.getString("default_route", "None")
        mCurrentWalkRoute = if (defaultRouteName == "None") {
            Route("No route set")
        } else {
            DBManager.getRoutes(context!!, defaultRouteName)[0]
        }
        if (routeNames.size > 0 && defaultRouteName == "None") {
            val adapter = ArrayAdapter<CharSequence>(context!!, R.layout.support_simple_spinner_dropdown_item)
            adapter.addAll(routeNames)
            adapter.add("None")
            val lpw = ListPopupWindow(context!!)
            lpw.setAdapter(adapter)
            lpw.anchorView = mFab
            lpw.setContentWidth(((mFab.parent as View).width / 2.5).toInt())
            lpw.setOnItemClickListener { adapterView, _, i, _ ->
                val selectedName = adapterView.getItemAtPosition(i).toString()
                if (selectedName != "None") {
                    mCurrentWalkRoute = DBManager.getRoutes(context!!, selectedName)[0]
                }
                startStopwatch()
                lpw.dismiss()
            }
            lpw.show()
        } else {
            startStopwatch()
        }
    }

    private fun startStopwatch() {
        mStartTime = SystemClock.elapsedRealtime()
        mTimeState = RUNNING
        updateUI()
        mHandler = Handler()
        mHandler!!.post(mRunnable)
    }

    internal fun pauseStopwatch() {
        mLastStopTime = SystemClock.elapsedRealtime()
        mTimeState = PAUSED
        updateUI()
        mHandler!!.removeCallbacks(mRunnable)
        buildNotification()
    }

    internal fun resumeStopwatch() {
        mStartTime += (SystemClock.elapsedRealtime() - mLastStopTime)
        mTimeState = RUNNING
        updateUI()
        mHandler!!.post(mRunnable)
    }

    internal fun stopWalk() {
        val walkTime = SystemClock.elapsedRealtime() - mStartTime
        val walk = Walk(walkTime, Date(), mCurrentWalkRoute!!)
        mLastStopTime = 0
        mCurrentWalkRoute = null
        mTimeState = STOPPED
        updateUI()
        DBManager.saveWalk(context!!, walk)
    }

    private fun updateUI() {
        when (mTimeState) {
            STOPPED -> {
                mFab.setImageResource(android.R.drawable.ic_media_play)
                mBtnR.isClickable = false
                mBtnR.visibility = View.INVISIBLE
                mRouteView.text = ""
                mProgressBar.progress = 0
                if (mNotificationManager.activeNotifications != null) mNotificationManager.cancel(NOTIFICATION_ID)
                if (mHandler != null) mHandler!!.removeCallbacks(mRunnable)
            }
            RUNNING -> {
                mFab.setImageResource(android.R.drawable.ic_media_pause)
                mBtnR.isClickable = true
                mBtnR.visibility = View.VISIBLE
                mRouteView.text = mCurrentWalkRoute!!.name
            }
            PAUSED -> {
                mFab.setImageResource(android.R.drawable.ic_media_play)
                mBtnR.isClickable = true
                mBtnR.visibility = View.VISIBLE
                mTimeView.text = DateFormat.format("mm:ss", Date(mLastStopTime - mStartTime))
                mRouteView.text = mCurrentWalkRoute!!.name
            }
        }
    }

    private fun buildNotification() {
        val intent = Intent(activity, ActivityMain::class.java).apply {
            flags = Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_SINGLE_TOP
        }
        val pendingIntent = PendingIntent.getActivity(activity, 0, intent, 0)
        val builder = NotificationCompat.Builder(activity!!, CHANNEL_ID)
                .setSmallIcon(R.drawable.ic_current)
                .setContentTitle(if (mTimeState == RUNNING) "Ongoing walk" else "Walk paused")
                .setSubText(mCurrentWalkRoute!!.name)
                .setContentText(DateFormat.format("mm:ss", Date(SystemClock.elapsedRealtime() - mStartTime)))
                .setOngoing(mTimeState == RUNNING)
                .setLargeIcon(BitmapFactory.decodeResource(resources, R.mipmap.ic_app))
                .setContentIntent(pendingIntent)
                .setAutoCancel(true)
        addNotificationActions(builder)
        mNotificationManager.notify(NOTIFICATION_ID, builder.build())
    }

    private fun addNotificationActions(builder: NotificationCompat.Builder) {
        val btn1Intent = Intent(activity, ActivityMain::class.java)
        val btn1PendingIntent = PendingIntent.getActivity(activity, 0, btn1Intent, 0)
        when (mTimeState) {
            RUNNING -> {
                btn1Intent.action = getString(R.string.notification_action_pause)
                builder.addAction(NotificationCompat.Action.Builder(
                        R.drawable.ic_walk,
                        getString(R.string.notification_action_pause),
                        btn1PendingIntent).build())
            }
            PAUSED -> {
                btn1Intent.action = getString(R.string.notification_action_resume)
                builder.addAction(NotificationCompat.Action.Builder(
                        R.drawable.ic_walk,
                        getString(R.string.notification_action_resume),
                        btn1PendingIntent).build())
            }
            else -> {
            }
        }
        //no checks as far as TimeState.STOPPED has no notification
        val stopIntent = Intent(activity, ActivityMain::class.java)
        stopIntent.action = getString(R.string.notification_action_stop)
        val stopPendingIntent = PendingIntent.getActivity(activity, 0, stopIntent, 0)
        builder.addAction(NotificationCompat.Action.Builder(
                R.drawable.ic_walk,
                getString(R.string.notification_action_stop),
                stopPendingIntent).build())
    }

    private fun setProgressSeconds(seconds: Int) {
        if (mAnimator != null && mAnimator!!.isRunning) mAnimator!!.cancel()
        mAnimator = ObjectAnimator.ofInt(mProgressBar, "progress", seconds * 100)
        mAnimator!!.duration = 1500
        mAnimator!!.interpolator = DecelerateInterpolator()
        mAnimator!!.start()
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        menu!!.clear()
        inflater!!.inflate(R.menu.menu_current, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        when (item!!.itemId) {
            R.id.action_settings -> startActivity(Intent(activity, ActivitySettings::class.java))
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onResume() {
        super.onResume()
        if (mTimeState == null) mTimeState = STOPPED
        else if (mTimeState == RUNNING) mHandler!!.post(mRunnable)
        updateUI()
    }

    private fun createNotificationChannel() {
        mNotificationManager = activity!!.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        mChannel = NotificationChannel(CHANNEL_ID, getString(R.string.channel_name), NotificationManager.IMPORTANCE_LOW)
        mChannel.description = getString(R.string.channel_description)
        mChannel.enableLights(mSharedPref.getBoolean("notifications_led", false))
        mChannel.lightColor = Color.GREEN
        mChannel.enableVibration(mSharedPref.getBoolean("notifications_vibrate", false))
        mChannel.vibrationPattern = longArrayOf(100, 200, 300, 400, 500, 400, 300, 200, 400)
        mNotificationManager.createNotificationChannel(mChannel)
    }

    private enum class TimeState {
        RUNNING, PAUSED, STOPPED
    }

    companion object {
        private const val NOTIFICATION_ID = 1
        private const val CHANNEL_ID = "channel_walk"
    }
}


